django/transformer/transform/dataframe.py

```diff
@@ -64,71 +65,66 @@ def clean_data(data, attributes: List[Attribute], primary_key_col: SqlColumn, pr
 
 
 def merge_by_attributes(data, attributes: List[Attribute], primary_key_value: str):
-    """Apply merging scripts.
-     Takes as input a dict of the form
-
+    """Takes as input a dict of the form
+    {
+        (attribute.path, input_group1.id, (table1, column1)): values,
+        (attribute.path, input_group1.id, (table2, column2)): values,
+        (attribute.path, input_group2.id, (table3, column3)): values,
+        (CONDITION_FLAG, (table4, column4)): values,
+        ...
+    }
+    and outputs
     {
-         (input_group1.id, (table1, column1)): val,
-         (input_group1.id, (table2, column2)): val,
-         (input_group2.id, (table3, column3)): val,
-         (CONDITION_FLAG, (table4, column4)): val,
-         ...
-     }
-
-     and outputs
-
-     {
-         attribute1.path: val,
-         attribute2.path: val,
-         ...
-     }
-
-     where values are merged with the mergig scripts.
+        attribute1.path: values,
+        attribute2.path: values,
+        ...
+    }
+    where values are merged with the mergig scripts.
     """
-    merged_data = {}
+    merged_data = defaultdict(list)
     for attribute in attributes:
-        for input_group in attribute.input_groups:
-            # Check conditions
-            if input_group.check_conditions(data):
-                cur_attr_columns = [value for key, value in data.items() if key[0] == input_group.id]
-
-                if not cur_attr_columns:
-                    if input_group.static_inputs:
-                        # If attribute is static, use static input
-                        # Otherwise, attribute is not a leaf or has no inputs
-                        if len(input_group.static_inputs) != 1:
-                            raise ValueError(
-                                f"The mapping contains an attribute ({attribute.path}) "
-                                "with several static inputs (and no sql input)"
+        data_for_attribute = {key: value for key, value in data.items() if key[0] == attribute.path}
+        nb_rows_for_attribute = max(len(col) for col in data_for_attribute.values()) if data_for_attribute else 1
+
+        for row_ind in range(nb_rows_for_attribute):
+            row_data = {col_key: get_element_in_array(col, row_ind) for col_key, col in data.items()}
+
+            no_group_matched = True
+            for input_group in attribute.input_groups:
+
+                if all(condition.check(row_data) for condition in input_group.conditions):
+                    cur_group_columns = [value for key, value in row_data.items() if key[1] == input_group.id]
```

Comment:
this is not easy to read (to say the least), would you be kind with regular humans like me and explain what's going on in plain english ?